core-processor/src/ext.rs

```diff
@@ -872,18 +876,16 @@ impl Ext {
         pages: WasmPage,
         mem: &mut impl Memory,
     ) -> Result<WasmPage, ProcessorAllocError> {
-        // Charge gas for memory grow.
-        // TODO: move charging for grow inside alloc function, so that we can skip refunding #2337
-        let gas_refunder =
-            GasRefunder::charge_if_enough(self, self.context.page_costs.mem_grow.calc(pages))?;
-        let AllocInfo { page, not_grown } =
-            self.context.allocations_context.alloc::<G>(pages, mem)?;
-
-        // Refund gas if memory has not been grown to `pages` size.
-        gas_refunder
-            .refund(self, self.context.page_costs.mem_grow.calc(not_grown))
-            .unwrap_or_else(|err| unreachable!("Alloc implementation error: {err}"))?;
-
+        let AllocInfo { page, .. } =
+            self.context
+                .allocations_context
+                .alloc::<G>(pages, mem, |pages| {
```

Comment:
I don't know whether it is better (if feasible) to implement a charger in `alloc` as a second generic instead of closure. In theory, static dispatch would be a little bit more efficient. But maybe it is not worth messing up.

Reply:
Isn't it a static dispatch? The `alloc` method takes impl FnOnce which is a generic, doesn't it? It does take a function pointer.

Reply:
This is what I've found in an earlier edition of the Book
https://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/first-edition/closures.html#:~:text=In%20many%20languages%2C%20closures%20are,often%20take%20closures%20as%20arguments.
Unfortunatelly, I can't find the same in the most recent edition.